### PR TITLE
More accelerate upstream

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
# What does this PR do?

This PR uses more of the upstream accelerate and only patches what still needs patching:

- Sequence parallelism (just realised we should add this to the list of features in accelerate gaudi docs)
- FP8 (minimize memory and TE_FusedAttention)
- Fast DDP

also removed foece_autocast from GaudiAccelerator since it's interchangeable with `mixed_precision == bf16`.
This will need slow testing to make sure none of these features was impacted !

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/accelerate/accelerator.py [hunk 1]
```diff
@@ -15,619 +15,78 @@
...
-                        from deepspeed.ops.adam import DeepSpeedCPUAdam
-
-                        defaults = {k: v for k, v in optimizer.defaults.items() if k in ["lr", "weight_decay"]}
-                        optimizer = DeepSpeedCPUAdam(optimizer.param_groups, **defaults)
-                    kwargs["optimizer"] = optimizer
-                    if scheduler is not None:
-                        if type(scheduler).__name__ in deepspeed.runtime.lr_schedules.VALID_LR_SCHEDULES:
-                            kwargs["lr_scheduler"] = scheduler
+        orig_num_models = len(self._models)
+        tuple = super()._prepare_deepspeed(*args)
```
Comment (yafshar) [inline optimum/habana/accelerate/accelerator.py:81]:
```suggestion
        prepared_deepspeed = super()._prepare_deepspeed(*args)
```

Please use prepared_deepspeed or result instead of tuple to avoid confusion with the built-in type

File: optimum/habana/accelerate/accelerator.py [hunk 2]
```diff
@@ -15,619 +15,78 @@
...
-                    "If you want to call `accelerator.backward()` referencing a new model/engine, "
-                    "please create a separate `Accelerator()` instance and call `accelerator.prepare()` on it."
-                )
-            self._models.append(engine)
-            if optimizer is not None:
-                self._optimizers.append(optimizer)
-            if scheduler is not None:
-                self._schedulers.append(scheduler)
-        return tuple(result)
+        return tuple
```
Comment (yafshar) [inline optimum/habana/accelerate/accelerator.py:87]:
```suggestion
        return prepared_deepspeed
```

same as above

File: optimum/habana/accelerate/accelerator.py (hunk 1 above)
Comment (IlyasMoutawwakil) [reply optimum/habana/accelerate/accelerator.py:81]: I tried to keep the nomenclature used in accelerate

**Comment by yafshar** [comment]
@IlyasMoutawwakil I will start testing this PR. I will close https://github.com/huggingface/optimum-habana/pull/1978 later after finishing some tests

**Comment by yafshar** [comment]
@IlyasMoutawwakil I am in the middle of testing. Till now one test is failing

```sh
>>> RUN_SLOW=true python -m pytest tests/test_encoder_decoder.py -s -v -k test_text_translation_bf16[t5-small-Habana/t5-2-1]

***** Running Prediction *****
[INFO|trainer.py:1923] 2025-05-15 20:16:24,969 >>   Num examples = 200
[INFO|trainer.py:1926] 2025-05-15 20:16:24,969 >>   Batch size = 2
Traceback (most recent call last):
  File "/root/optimum-habana/examples/translation/run_translation.py", line 728, in <module>
    main()
  File "/root/optimum-habana/examples/translation/run_translation.py", line 682, in main
    predict_results = trainer.predict(
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/trainer_seq2seq.py", line 256, in predict
    return super().predict(test_dataset, ignore_keys=ignore_keys, metric_key_prefix=metric_key_prefix)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/trainer.py", line 1826, in predict
    output = eval_loop(
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/trainer.py", line 1977, in evaluation_loop
    losses, logits, labels = self.prediction_step(model, inputs, prediction_loss_only, ignore_keys=ignore_keys)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/trainer_seq2seq.py", line 350, in prediction_step
    raise error
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/trainer_seq2seq.py", line 340, in prediction_step
    generated_tokens = self.model.generate(
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/generation/utils.py", line 1690, in generate
    result = self._sample(
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/generation/utils.py", line 2662, in _sample
    outputs = self(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1742, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1753, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/graphs.py", line 745, in forward
    return wrapped_hpugraph_forward(
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/graphs.py", line 610, in wrapped_hpugraph_forward
    outputs = orig_fwd(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/utils/operations.py", line 818, in forward
    return model_forward(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/utils/operations.py", line 806, in __call__
    return convert_to_fp32(self.model_forward(*args, **kwargs))
  File "/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py", line 44, in decorate_autocast
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/models/t5/modeling_t5.py", line 556, in gaudi_T5ForConditionalGeneration_forward
    decoder_outputs = self.decoder(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1742, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1848, in _call_impl
    return inner()
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1796, in inner
    result = forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/models/t5/modeling_t5.py", line 432, in gaudi_T5Stack_forward
    layer_outputs = layer_module(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1742, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1848, in _call_impl
    return inner()
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1796, in inner
    result = forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/models/t5/modeling_t5.py", line 254, in gaudi_T5Block_forward
    self_attention_outputs = self.layer[0](
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1742, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1848, in _call_impl
    return inner()
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1796, in inner
    result = forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/models/t5/modeling_t5.py", line 204, in gaudi_T5LayerSelfAttention_forward
    attention_output = self.SelfAttention(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1742, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1848, in _call_impl
    return inner()
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1796, in inner
    result = forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/models/t5/modeling_t5.py", line 124, in gaudi_T5Attention_forward
    key_states = project(
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/models/t5/modeling_t5.py", line 108, in project
    hidden_states = past_key_value.index_copy_(-2, token_idx - 1, hidden_states)
RuntimeError: cpu fallback is not supported during hpu graph capturing. You should run inference in lazy mode only with `use_lazy_mode=True` and `use_hpu_graphs_for_inference=False`.
FAILED
```

I have not checked the issue yet

File: optimum/habana/accelerate/accelerator.py [hunk 3]
```diff
@@ -15,619 +15,78 @@
...
-            and os.environ.get("ACCELERATE_BYPASS_DEVICE_MAP", "false") != "true"
-        ):
-            raise ValueError(
-                "You can't train a model that has been loaded with `device_map='auto'` in any distributed mode."
-                " Please rerun your script specifying `--num_processes=1` or by launching with `python {{myscript.py}}`."
-            )
-
-        # The following block is executed only when force_autocast is True
-        # because forward+backward+loss is already wrapped with autocast in Trainer
-        if self.native_amp and self.force_autocast:
```
Comment (yafshar) [inline optimum/habana/accelerate/accelerator.py:190]: @IlyasMoutawwakil removing this part causes issue!

in the upstream `self.native_amp` is True when using bf16, and it will enter the conditional https://github.com/huggingface/accelerate/blob/417bc529654a70e61013fd21263826a2f1f9e1a6/src/accelerate/accelerator.py#L1507

For Gaudi it should not do that because forward+backward+loss is already wrapped with autocast in Trainer

Comment (yafshar) [reply optimum/habana/accelerate/accelerator.py:190]: The fail still happening. We have more than 7 fails till now. For the above case if I disable the conditional in accelerate. the test is passing!

Comment (yafshar) [reply optimum/habana/accelerate/accelerator.py:190]: For the above issue we can solve it by 

```sh
diff --git a/optimum/habana/transformers/models/t5/modeling_t5.py b/optimum/habana/transformers/models/t5/modeling_t5.py
index ff968612..e3007d4a 100644
--- a/optimum/habana/transformers/models/t5/modeling_t5.py
+++ b/optimum/habana/transformers/models/t5/modeling_t5.py
@@ -103,9 +103,9 @@ def gaudi_T5Attention_forward(
                 # self-attn
                 # (batch_size, n_heads, key_length, dim_per_head)
                 if token_idx is None:
-                    hidden_states = torch.cat([past_key_value, hidden_states], dim=2)
+                    hidden_states = torch.cat([past_key_value.to(hidden_states.dtype), hidden_states], dim=2)
                 else:
-                    hidden_states = past_key_value.index_copy_(-2, token_idx - 1, hidden_states)
+                    hidden_states = past_key_value.to(hidden_states.dtype).index_copy_(-2, token_idx - 1, hidden_states)
             elif past_key_value.shape[2] != key_value_states.shape[1]:
                 # checking that the `sequence_length` of the `past_key_value` is the same as
                 # the provided `key_value_states` to support prefix tuning
```

@regisss do you have any suggestion?

Comment (regisss) [reply optimum/habana/accelerate/accelerator.py:190]: Ideally the trainer should not manage that, but I think there was an issue with `convert_outputs_to_fp32`

**Comment by yafshar** [comment]
I am doing a fresh round of tests

**Comment by yafshar** [comment]
@IlyasMoutawwakil can you sync this PR with main

**Comment by regisss** [comment]
> @IlyasMoutawwakil can you sync this PR with main

Done

File: optimum/habana/accelerate/accelerator.py [hunk 4]
```diff
@@ -15,619 +15,80 @@
 
 from __future__ import annotations
 
-import functools
-import math
-import os
-from dataclasses import make_dataclass
-from types import MethodType
-
+import accelerate
```
Comment (yafshar) [inline optimum/habana/accelerate/accelerator.py:18]:
```suggestion
import os

import accelerate
```

`os` is required again since it is added from main in `prepare_data_loader`

**Comment by IlyasMoutawwakil** [comment]
@emascarenhas done !

File: optimum/habana/accelerate/state.py
Comment (regisss) [inline optimum/habana/accelerate/state.py:1]: Empty file so let's delete it?

Comment (IlyasMoutawwakil) [reply optimum/habana/accelerate/state.py:1]: it is deleted, I was also confused by the "Empty file." description but apparently it was empty and I removed it in this PR.

**Comment by IlyasMoutawwakil** [comment]
@yafshar do you have some bandwidth for some testing of this branch ?

**Comment by yafshar** [comment]
> @yafshar do you have some bandwidth for some testing of this branch ?
Sure, I’ll take a look.

File: optimum/habana/accelerate/accelerator.py [hunk 5]
```diff
@@ -82,563 +43,49 @@
...
-        # it will be added in upstream accelerate
-        self.fp8_enabled = self.mixed_precision == "fp8" or mixed_precision == "fp8"
-
-        # will be fixed in upstream accelerate
-        self.has_fp8_handler = self.te_recipe_handler is not None or self.fp8_recipe_handler is not None
+        self.mpu = parallel_state
+        self.force_autocast = force_autocast
+        self.distribution_strategy = distribution_strategy
+        self.compiled_autograd_enabled = compiled_autograd_enabled
+        self.native_amp = self.native_amp and self.force_autocast
```
Comment (yafshar) [inline optimum/habana/accelerate/accelerator.py]:
```suggestion
        self.distribution_strategy = distribution_strategy
        self.compiled_autograd_enabled = compiled_autograd_enabled
        self.native_amp = self.native_amp and force_autocast
```

I do not think we need to store `force_autocast`!

Comment (IlyasMoutawwakil) [reply optimum/habana/accelerate/accelerator.py]: yeh ofc

**Comment by IlyasMoutawwakil** [comment]
@yafshar any performance regressions in your tests ? cuz we want to merge this PR

**Review by regisss — APPROVED** [review]
LGTM!
